RHIDP-1917: RBAC conditional policies in RHDH (#252)
redhat-developer/red-hat-developers-documentation-rhdh: modules/admin/ref-rbac-conditional-policy-definition.adoc

[id='ref-rbac-conditional-policy-definition_{context}']
= Conditional policies definition

The {product} provides API endpoints for conditional policies that you can use. For example, to retrieve the available conditional rules, which can help you define these policies, you can access the `GET [api/plugins/condition-rules]` endpoint.

The `api/plugins/condition-rules` returns the condition parameters schemas:

[source,json]
----
[
   {
      "pluginId": "catalog",
      "rules": [
         {

>             "name": "HAS_ANNOTATION",

Author: This isn't the file name; it's an example rule that has been defined.

               "required": [
                  "kinds"
               ],
               "additionalProperties": false,
               "$schema": "http://json-schema.org/draft-07/schema#"
            }
         },
         {
            "name": "IS_ENTITY_OWNER",
            "description": "Allow entities owned by a specified claim",
            "resourceType": "catalog-entity",
            "paramsSchema": {
               "type": "object",
               "properties": {
                  "claims": {
                     "type": "array",
                     "items": {
                        "type": "string"
                     },
                     "description": "List of claims to match at least one on within ownedBy"
                  }
               },
               "required": [
                  "claims"
               ],
               "additionalProperties": false,
               "$schema": "http://json-schema.org/draft-07/schema#"
            }
         }
      ]
   }
   ... <another plugin condition parameter schemas>
]
----

The RBAC backend API constructs a condition JSON object based on the previous condition schema.

== Examples of conditional policies

> In {product}, you can define conditional policies with or without criteria, such as:

Author: I've added:
You can use the following examples to define the conditions based on your use case.

Since conditions are not configured but defined.

You can apply the previous example condition to the RBAC REST API by adding additional parameters as follows:

[source,json]
----
{
  "result": "CONDITIONAL",
  "roleEntityRef": "role:default/test",
  "pluginId": "catalog",
  "resourceType": "catalog-entity",
  "permissionMapping": ["read"],
  "conditions": {
    "rule": "IS_ENTITY_OWNER",
    "resourceType": "catalog-entity",
    "params": {
      "claims": ["group:default/team-a"]
    }
  }
}
----
--

A condition with criteria::
+
--
Consider a condition with criteria, which displays catalogs only if user is a member of owner group OR displays list of all catalog user groups.

To add the criteria, you can add another rule as `IS_ENTITY_KIND` in the condition as follows:

.Example condition with criteria
[source,json]
----
{
  "anyOf": [ 
    {
      "rule": "IS_ENTITY_OWNER",
      "resourceType": "catalog-entity",
      "params": {
        "claims": ["group:default/team-a"]
      }
    },
    {
      "rule": "IS_ENTITY_KIND",
      "resourceType": "catalog-entity",
      "params": {
        "kinds": ["Group"]
      }
    }
  ]
}
----

[NOTE]
====

> Running conditions in parallel during creation is not supported. Therefore, consider defining separate condition requests or nesting your conditions based on the available criteria.

Reviewer: ```suggestion
Running conditions in parallel during creation is not supported. Therefore, consider defining nested conditional policies based on the available criteria.
```

I ended up getting some clarification from @AndrienkoAleksandr and it seems like we do not support separate conditional requests. So, I need to update the documentation in our plugin as well.

Author: Done.

[source,json]
----
{
  "result": "CONDITIONAL",
  "roleEntityRef": "role:default/test",
  "pluginId": "catalog",
  "resourceType": "catalog-entity",
  "permissionMapping": ["read"],
  "conditions": {
    "anyOf": [
      {
        "rule": "IS_ENTITY_OWNER",
        "resourceType": "catalog-entity",
        "params": {
          "claims": ["group:default/team-a"]
        }
      },
      {
        "rule": "IS_ENTITY_KIND",
        "resourceType": "catalog-entity",
        "params": {
          "kinds": ["Group"]
        }
      }
    ]
  }
}
----
--

The following examples can be used with {product-short} plugins. These examples can help you determine how to define conditional policies:

.Conditional policy defined for Keycloak plugin
[source,json]
----
{
  "result": "CONDITIONAL",
  "roleEntityRef": "role:default/developer",
  "pluginId": "catalog",
  "resourceType": "catalog-entity",
  "permissionMapping": ["update", "delete"],
  "conditions": {
    "not": {
      "rule": "HAS_ANNOTATION",
      "resourceType": "catalog-entity",
      "params": { "annotation": "keycloak.org/realm", "value": "<YOUR_REALM>" }
    }
  }
}
----

> The previous example of Keycloak plugin prevents users in the `role:default/developer` from updating or deleting users that are ingested unto the catalog from the Keycloak plugin.

Reviewer 2: suggestion: 'ingested unto' - > 'ingested into'

Author: Thanks, Done
